content/cost-optimization/docs/cost-optimization.md

```diff
@@ -0,0 +1,238 @@
+  containers:
+  - name: spark-kubernetes-executor
+```
+Node affinity can also be used here, this allows for more flexibility for the constraints defined. We recommend to use ‘hard affinity’ as highlighted in the code below for this purpose. For jobs which have strict SLA and are not suitable to run on Spot we suggest to use NoSchedule taint effect to ensure no Pods are scheduled. The key thing to note here is that the bulk of the compute required in a Spark job runs on executors and if they can be run on EC2 Spot instances you can benefit from the steep discount available with Spot instances.
+
+```YAML
+apiVersion: v1
+kind: Pod
+metadata:
+  labels:
+    spark-role: driver
+  namespace: emr-eks-workshop-namespace
+spec:
+  affinity: 
+      nodeAffinity: 
+          requiredDuringSchedulingIgnoredDuringExecution: 
+            nodeSelectorTerms: 
+            - matchExpressions: 
+              - key: 'eks.amazonaws.com/capacityType' 
+                operator: In 
+                values: 
+                - ON_DEMAND
+```
+
+```YAML
+apiVersion: v1
+kind: Pod
+metadata:
+  labels:
+    spark-role: executor
+  namespace: emr-eks-workshop-namespace
+spec:
+  affinity: 
+      nodeAffinity: 
+          requiredDuringSchedulingIgnoredDuringExecution: 
+            nodeSelectorTerms: 
+            - matchExpressions: 
+              - key: 'eks.amazonaws.com/capacityType' 
+                operator: In 
+                values: 
+                - SPOT
+```
+
+When Spot instances are interrupted the executors running on them may lose (if any) the shuffle and cached RDDs which would require re-computation. This requires more compute cycles to be spent which will impact the overall SLA of the EMR on EKS jobs. EMR on EKS has incorporated two new Spark features which can help to address these issues. In the following sections we will discuss them.
+
+**Node Decommissioning:**
+
+Node decommissioning is a Spark feature that enables the removal of an executor gracefully, by preserving its state before removing it and not scheduling any new jobs on it. This feature is particularly useful when the Spark executors are running on Spot instances, and the Spark executor node is interrupted via a ‘rebalance recommendation’ or ‘instance termination’ notice to reclaim the instance. 
+
+Node decommission begins when a Spark executor node receives a Spot Interruption Notice or Spot Rebalance Recommendation signal. The executor node immediately starts the process of decommissioning by sending a message to the Spark driver. The driver will identify the RDD/Shuffle files that it needs to migrate off the executor node in question, and will try to identify another Executor node which can take over the execution. If an executor is identified, the RDD/Shuffle files are copied to the new executor and the job execution continues on the new executor. If all the executors are busy, the RDD/Shuffle files are copied to an external storage.
```

Comment:
> If all the executors are busy, the RDD/Shuffle files are copied to an external storage.

To be specific, are we talking about `spark.storage.decommission.fallbackStorage.path` config? Last I had seen there was an issue in this feature. We were seeing the error `java.io.IOException: Stream is corrupted, net.jpountz.lz4.LZ4Exception: Error decoding offset <offset_id> of input buffer`

Reply:
Yes this is working now. I will update the exact version after some testing.